test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java

```diff
@@ -53,52 +57,143 @@
     private static final String VT2_NAME = "vt2";
     private static final String VT3_NAME = "vt3";
 
-    private static class WritableVirtualTable extends AbstractVirtualTable
+    private static class WritableVirtualTable extends AbstractWritableVirtualTable.SimpleWritableVirtualTable
     {
-        private final ColumnMetadata valueColumn;
-        private final Map<String, Integer> backingMap = new HashMap<>();
+        // <pk1, pk2> -> c1 -> c2 -> <v1, v2>
+        private final Map<Pair<String, String>, SortedMap<String, SortedMap<String, Pair<Integer, Long>>>> backingMap = new ConcurrentHashMap<>();
 
         WritableVirtualTable(String keyspaceName, String tableName)
         {
             super(TableMetadata.builder(keyspaceName, tableName)
                                .kind(TableMetadata.Kind.VIRTUAL)
-                               .addPartitionKeyColumn("key", UTF8Type.instance)
-                               .addRegularColumn("value", Int32Type.instance)
+                               .addPartitionKeyColumn("pk1", UTF8Type.instance)
+                               .addPartitionKeyColumn("pk2", UTF8Type.instance)
+                               .addClusteringColumn("c1", UTF8Type.instance)
+                               .addClusteringColumn("c2", UTF8Type.instance)
+                               .addRegularColumn("v1", Int32Type.instance)
+                               .addRegularColumn("v2", LongType.instance)
                                .build());
-            valueColumn = metadata().regularColumns().getSimple(0);
         }
 
         @Override
         public DataSet data()
         {
             SimpleDataSet data = new SimpleDataSet(metadata());
-            backingMap.forEach((key, value) -> data.row(key).column("value", value));
+            backingMap.forEach((pkPair, c1Map) ->
+                    c1Map.forEach((c1, c2Map) ->
+                    c2Map.forEach((c2, valuePair) -> data.row(pkPair.left, pkPair.right, c1, c2)
+                            .column("v1", valuePair.left)
+                            .column("v2", valuePair.right))));
             return data;
         }
 
         @Override
-        public void apply(PartitionUpdate update)
+        protected void applyPartitionDelete(Object[] partitionKeyColumnValues)
+        {
+            String pk1 = (String) partitionKeyColumnValues[0];
+            String pk2 = (String) partitionKeyColumnValues[1];
+            backingMap.remove(Pair.create(pk1, pk2));
+        }
+
+        @Override
+        protected void applyRangeTombstone(Object[] partitionKeyColumnValues,
```

Comment:
I believe for most writable VTs it won't be required to support range deletes. However, it is possible as it is demonstrated in this method.